File: backend/app/schemas/protocol.py
"""标准化双任务测评流程 API schemas。"""
from datetime import datetime

from pydantic import Field, field_validator
from app.schemas.base import ApiModel as BaseModel


class ProtocolTaskOut(BaseModel):
    id: str
    title: str
    description: str
    scenario: str
    estimated_minutes: int
    protocol_order: int
    stimulus_data: dict | None = None


class ProtocolQuestionnaireItemOut(BaseModel):
    id: str
    dimension: str
    text: str
    scale_min: int
    scale_max: int
    display_order: int


class ProtocolNarrationAssetOut(BaseModel):
    id: str
    slot_key: str
    version: int
    original_filename: str
    mime_type: str
    size_bytes: int
    created_at: datetime


class AssessmentProtocolOut(BaseModel):
    version: str
    questionnaire_enabled: bool = False
    questionnaire_source: str
    task_order_code: str = "AB"
    order_source: str = "default"
    tasks: list[ProtocolTaskOut]
    questionnaire_items: list[ProtocolQuestionnaireItemOut]
    likert_labels: dict[int, str]
    narration_assets: list[ProtocolNarrationAssetOut] = Field(default_factory=list)


class RunCreateIn(BaseModel):
    consent: bool


class RunSessionOut(BaseModel):
    id: str
    task_id: str
    sequence_no: int
    status: str


class QuestionnaireAnswerOut(BaseModel):
    item_id: str
    value: int


class AssessmentRunOut(BaseModel):
    id: str
    user_id: str
    status: str
    current_stage: str
    protocol_version: str
    questionnaire_enabled: bool = True
    questionnaire_source: str
    task_order_code: str = "AB"
    order_assignment_id: str | None = None
    consented_at: datetime
    started_at: datetime
    completed_at: datetime | None = None
    sessions: list[RunSessionOut]
    questionnaire_answers: list[QuestionnaireAnswerOut] = []
    questionnaire_participant_name: str | None = None


class RunStageIn(BaseModel):
    stage: str = Field(
        pattern="^(device_check|instructions|practice|task_1|task_2|questionnaire|review)$"
    )


class QuestionnaireAnswerIn(BaseModel):
    item_id: str = Field(min_length=1, max_length=36)
    value: int = Field(ge=1, le=7)


class QuestionnaireSubmitIn(BaseModel):
    answers: list[QuestionnaireAnswerIn] = Field(min_length=1, max_length=100)
    participant_name: str = Field(min_length=1, max_length=255)

    @field_validator("participant_name")
    @classmethod
    def normalize_participant_name(cls, value: str) -> str:
        normalized = " ".join(value.split())
        if not normalized:
            raise ValueError("请填写您的姓名或参加实验时使用的微信名等标识")
        return normalized


class ProtocolConfigUpdate(BaseModel):
    questionnaire_enabled: bool


class ProtocolConfigOut(BaseModel):
    questionnaire_enabled: bool
    updated_at: datetime | None = None


class NarrationAssetOut(BaseModel):
    id: str
    slot_key: str
    label: str
    source_text: str
    original_filename: str
    mime_type: str
    size_bytes: int
    sha256: str
    version: int
    is_active: bool
    uploaded_by: str
    created_at: datetime

    model_config = {"from_attributes": True}


class NarrationSlotOut(BaseModel):
    slot_key: str
    label: str
    source_text: str
    category: str
    asset: NarrationAssetOut | None = None


class TaskOrderAssignmentIn(BaseModel):
    ordered_task_ids: list[str] = Field(min_length=2, max_length=2)


class TaskOrderBalanceIn(BaseModel):
    user_ids: list[str] = Field(min_length=1, max_length=500)


class TaskOrderStudentOut(BaseModel):
    user_id: str
    username: str
    name: str
    class_group: str | None
    ordered_task_ids: list[str]
    order_code: str
    assigned_by: str | None = None
    assigned_at: datetime | None = None
    has_in_progress_run: bool = False


class TaskOrderOverviewOut(BaseModel):
    tasks: list[ProtocolTaskOut]
    students: list[TaskOrderStudentOut]
    total: int = 0
    page: int = 1
    page_size: int = 50

File: frontend/src/api/protocol.ts
import apiClient from './client'

export interface ProtocolTask {
  id: string
  title: string
  description: string
  scenario: string
  estimated_minutes: number
  protocol_order: number
  stimulus_data: {
    type?: 'scatter' | 'athletes'
    target?: { x: number; y: number }
    points?: Array<{ x: number; y: number }>
    athletes?: Array<Record<string, string | number>>
    columns?: Array<{ key: string; label: string }>
    question?: string
    stimulus_version?: string
    image_path?: string
    image_title?: string
    image_sha256?: string
  } | null
}

export interface ProtocolQuestionnaireItem {
  id: string
  dimension: string
  text: string
  scale_min: number
  scale_max: number
  display_order: number
}

export interface ProtocolNarrationAsset {
  id: string
  slot_key: string
  version: number
  original_filename: string
  mime_type: string
  size_bytes: number
  created_at: string
}

export interface AssessmentProtocol {
  version: string
  questionnaire_enabled: boolean
  questionnaire_source: string
  task_order_code: string
  order_source: 'default' | 'assignment' | 'active_run'
  tasks: ProtocolTask[]
  questionnaire_items: ProtocolQuestionnaireItem[]
  likert_labels: Record<string, string>
  narration_assets: ProtocolNarrationAsset[]
}

export interface RunSession {
  id: string
  task_id: string
  sequence_no: number
  status: string
}

export interface AssessmentRun {
  id: string
  user_id: string
  status: string
  current_stage: string
  protocol_version: string
  questionnaire_enabled: boolean
  questionnaire_source: string
  task_order_code: string
  order_assignment_id: string | null
  consented_at: string
  started_at: string
  completed_at: string | null
  sessions: RunSession[]
  questionnaire_answers: Array<{ item_id: string; value: number }>
  questionnaire_participant_name: string | null
}

export const protocolApi = {
  getProtocol() {
    return apiClient.get<AssessmentProtocol>('/assessment/protocol')
  },

  createRun(consent: boolean) {
    return apiClient.post<AssessmentRun>('/assessment/runs', { consent })
  },

  getCurrentRun() {
    return apiClient.get<AssessmentRun | null>('/assessment/runs/current')
  },

  advanceStage(runId: string, stage: string) {
    return apiClient.patch<AssessmentRun>(`/assessment/runs/${runId}/stage`, { stage })
  },

  submitQuestionnaire(
    runId: string,
    answers: Array<{ item_id: string; value: number }>,
    participantName: string
  ) {
    return apiClient.post<AssessmentRun>(
      `/assessment/runs/${runId}/questionnaire`,
      { answers, participant_name: participantName }
    )
  },

  completeRun(runId: string) {
    return apiClient.post<AssessmentRun>(`/assessment/runs/${runId}/complete`)
  },

  getNarrationAudio(assetId: string) {
    return apiClient.get<Blob>(`/narrations/${assetId}/audio`, {
      responseType: 'blob'
    })
  }
}
